src/server.c

```diff
@@ -5909,6 +5909,57 @@ void sendChildInfo(childInfoType info_type, size_t keys, char *pname) {
     sendChildInfoGeneric(info_type, keys, -1, pname);
 }
 
+/* Use 'MADV_DONTNEED' to release memory to operation system quickly. */
+void redisUnusedMemory(void *ptr) {
+    /* madvise(MADV_DONTNEED) may not work if enabled Transparent Huge Pages. */
+    if (server.thp_enabled == 1) return;
+    if (ptr == NULL) return;
+
+    size_t real_size = zmalloc_size(ptr);
+    if (real_size < server.page_size) return;
+
+    size_t page_size_mask = server.page_size - 1;
+    void *real_ptr = zmalloc_realptr(ptr);
+    char *aligned_ptr = (char *)((size_t)(real_ptr) & ~page_size_mask);
+    long aligned_size = real_size - ((char*)real_ptr-aligned_ptr);
+    if (aligned_size >= (long)server.page_size) {
+        madvise((void *)aligned_ptr, aligned_size&~page_size_mask, MADV_DONTNEED);
+    }
+}
+
+void unusedClientMemory(client *c) {
+    listIter li;
+    listNode *ln;
+    listRewind(c->reply, &li);
+    while((ln = listNext(&li))) {
+        void *bulk = listNodeValue(ln);
+        /* Default bulk size is 16k, actually it has extra data, maybe it
+         * occupies 20k according to jemalloc bin size if using jemalloc. */
+        if (bulk) redisUnusedMemory(bulk);
+    }
+
+    /* The 'client' is a bit big because it has 16k default reply buffer. */
+    redisUnusedMemory(c);
+}
+
+/* In child process, we don't need some memory, but it is much possible
+ * to change them because clients' heavy requests. We can free them ASAP. */
+void redisUnusedMemoryInChild(void) {
+    /* Replication backlog. */
+    if (server.repl_backlog != NULL) {
+        redisUnusedMemory(server.repl_backlog);
+    }
+
+    /* All clients memory. */
+    listIter li;
+    listNode *ln;
+    listRewind(server.clients, &li);
+    while((ln = listNext(&li))) {
+        client *c = listNodeValue(ln);
+        unusedClientMemory(c);
+    }
```

Comment:
my benchmarks showed that if there's no write traffic on the parent, the extra calls for madvise don't slow down the saving. but this loop might. sometimes the server may have many thousands of clients.
we need to re-think this.

Reply:
Yes, we should avoid too much loop, i am not sure, but clients' memory will be changed if there is much write traffic.